### ISSUE TITLE
MemLeak

### ISSUE THREAD

**Issue by haquocviet** [google/benchmark#782]
I constantly see memory leak of total 356 bytes occurring when I run the benchmark.
Digging into source I found this line in sysinfo.cc file at line 577:
static const CPUInfo* info = new CPUInfo();

Is this a mistake or an intentionally made? 
If it's not a mistake so what is reason behind it to tolerate the leak?

**Comment by EricWF** [comment]
The leak is intentional and  it sucks your tooling is reporting it :-( 
Are you using valgrind? Perhaps we can suppress it somehow?


During program termination users could still be touching the `CPUInfo`, so we need to be sure its destructor hasn't run.  Using `new` is a common pattern to achieve this.

**Comment by haquocviet** [comment]
I use it in MSVC. 
Regardless which tool is used, I think the leak indeed exists because the CPUInfo object is allocated but never be freed anywhere.
Though that leak is not a disaster in the library but any careful programmer could be irritated when look at the leak report.

**Comment by EricWF** [comment]
You're right. The object is never freed. Because we can never be sure it is safe to do so. [For example]( https://wandbox.org/permlink/JHc3FydzO2nELvNP)

Let's figure out how to suppress the diagnostic false positive.

**Comment by haquocviet** [comment]
If the situation in the benchmark is similar to the example you cited then there should be a design flaw since it lets tasks still continuing after the program's main function goes out of scope.

**Comment by LebedevRI** [comment]
I agree with @EricWF, i'd say this is working as expected.
It's very common pattern for such static const global objects.
While sure, this is a leak in the sense that the memory is not freed by the program before exit,
it will be immediately freed by the system the very next moment. So it's rather harmless..

**Comment by brenoguim** [comment]
It could be seen as harmlful in the sense that it hides potential bugs. If no one should be using the object when shutdown starts, then having the proper destructor would point bugs.

Now, if it is intended by design that users would/could continue to use this object even during shutdown, then seems to be a fine solution. I don't see how that feature could be useful though.

**Comment by dominichamon** [comment]
It's entirely possible (unlikely i think) that a user has a CPUInfo in some statically allocated memory in their binary. If we were to run the destructor for the CPUInfo and their destructor accessed that CPUInfo then we'd have a crash on shutdown (due to the destructor order issue).

Initialization is thread-safe (in C++11, which we require in the library), and we avoid any destruction issues by not running the destructor.

This is the standard approach to safely managing static objects.

**Comment by natemc** [comment]
May I suggest the [Schwarz (aka nifty) counter](https://en.wikibooks.org/wiki/More_C%2B%2B_Idioms/Nifty_Counter) instead of leaking this object? I'm new to google benchmark and don't know how CPUInfo is used, so maybe that solution won't work in this case, but that's how the standard library safely knows when to destruct global objects (e.g., std::cout).

I was trying out google benchmark on some code that overrides the global operator new to invoke a custom memory allocator, and the custom allocator complained about this leak.  Ignoring this leak could hide a bug in my code.

**Comment by dominichamon** [comment]
Please see the earlier comments in this thread. It's intentional to not free the object. The Schwarz counter is a fine construct, but as the CPUInfo object has a trivial destructor its execution isn't subject to ordering. As such, there's no need to have the construct in place, we just don't need to run it at all.

We're currently following the common pattern from the [Google C++ style guide](https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables) as we have a non-trivial (non-constexpr) constructor.

I suggest that you might want to have your custom allocator ignore these global "leaks" as they're not program-affecting.
